cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeEngineCredentials.java

```diff
@@ -0,0 +1,141 @@
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.common.io.CharStreams;
+import com.google.gson.Gson;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Date;
+
+/**
+ * Provides ComputeEngineCredentials either locally if no endpoint is defined, or remotely if endpoint is defined.
+ */
+public final class ComputeEngineCredentials extends GoogleCredentials {
+  private static final Logger LOG = LoggerFactory.getLogger(ComputeEngineCredentials.class);
+  private static final Gson GSON = new Gson();
+  /**
+   * Time (in millisecond) to refresh the credentials before it expires.
+   */
+  private static final int NUMBER_OF_RETRIES = 10;
+  private static final int MIN_WAIT_TIME = 5000;
+  private static final int MAX_WAIT_TIME = 10000;
+  private static ComputeEngineCredentials computeEngineCredentials;
+  private static final byte[] lock = new byte[0];
+
+  private final String endPoint;
+
+  private ComputeEngineCredentials(String endPoint) {
+    this.endPoint = endPoint;
+  }
+
+  /**
+   * Returns ComputeEngineCredentials, and initializes it if doesn't exist.
+   * @param endPoint optional endpoint to be used for fetching GoogleCredentials remotely.
+   * @return ComputeEngineCredentials
+   */
+  public static ComputeEngineCredentials createOrGet(String endPoint) {
+    if (computeEngineCredentials == null) {
+      synchronized (lock) {
+        if (computeEngineCredentials == null) {
+          computeEngineCredentials = new ComputeEngineCredentials(endPoint);
+        }
+      }
+    }
+
+    return computeEngineCredentials;
+  }
+
+  private static AccessToken getAccessToken(String endPoint) throws IOException {
+    if (endPoint != null) {
+      return getAccessTokenRemotely(endPoint);
+    }
+
+    return getAccessTokenLocally();
+  }
+
+  private static AccessToken getAccessTokenLocally() throws IOException {
+    try {
+      GoogleCredentials googleCredentials = com.google.auth.oauth2.ComputeEngineCredentials.create();
+      return googleCredentials.refreshAccessToken();
+    } catch (IOException e) {
+      throw new IOException("Unable to get credentials from the environment. "
+                              + "Please explicitly set the account key.", e);
+    }
+  }
+
+  private static AccessToken getAccessTokenRemotely(String endPoint) throws IOException {
+    ExponentialBackOff backOff = new ExponentialBackOff.Builder()
+      .setInitialIntervalMillis(MIN_WAIT_TIME)
+      .setMaxIntervalMillis(MAX_WAIT_TIME).build();
+
+    for (int i = 0; i < NUMBER_OF_RETRIES; i++) {
+      try {
+        return fetchAccessToken(endPoint);
+      } catch (Exception ex) {
+        LOG.warn("Failed to fetch GoogleCredentials. Will retry.", ex);
```

Comment:
Maybe unnecessary to log intermediate failures, otherwise could be too spammy.

Reply:
Agree.